### PR TITLE
AWS Auth v4 Part 1: Authorization header auth without payload sha256 verification

### PR DESCRIPTION
**Pull request by shino**
This PR is the first part to implement AWS version 4 authentication scheme. [1]
Subsequent tasks are #1059 #1058 #1060 .

This PR includes:
- Implement authorization header / non- `aws-chunked` case.
  Payload sha256 verification is _out of scope_ (see #1059 )
- Test v4 Authorization header cases by boto_test
- Keep boto tests of v2 auth along with v4
- Kill switch of v4 auth and disabled by default

[1] Authenticating Requests (AWS Signature Version 4) - Amazon Simple Storage Service
http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html

### CODE REVIEW

**Comment by Basho-JIRA** [comment]
testing, testing... 

***[posted via JIRA by Shunichi Shinohara]***

**Comment by kuenishi** [comment]
I would recommend squashing several fix to make it comprehensive and bisectable.

**Comment by shino** [comment]
Did some interactive rebase and force pushed :fire:

File: src/riak_cs_s3_auth.erl [hunk 1]
```diff
@@ -77,6 +79,18 @@ authenticate(User, Signature, RD, _Ctx) ->
...
+%% Currently support signature v2 only, does NOT support signature v4.
+identify_by_query_string(RD) ->
+    {wrq:get_qs_value(?QS_KEYID, RD), wrq:get_qs_value(?QS_SIGNATURE, RD)}.
+
+parse_auth_header("AWS4-HMAC-SHA256 " ++ String) ->
+    case riak_cs_config:auth_v4_enabled() of
+        true ->
+            parse_auth_v4_header(String, undefined, []);
+        _ ->
+            {failed, {auth_not_supported, "AWS4-HMAC-SHA256"}}
```
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:97]: ~~I'd recommend `s/supported/enabled/`.~~

Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:97]: Ah, this message comes out to response to clients. not supported is correct message as a whole.

File: src/riak_cs_s3_auth.erl [hunk 2]
```diff
@@ -43,14 +43,16 @@
 identify(RD,_Ctx) ->
```
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:47]: The spec have changed to 

``` erlang
-spec identify(#reqdata{}, #context{}) -> {string() | undefined, string() | {v4, proplists:proplist()}.
```

 ? Can we narrow down from the behaviour callback definition?

File: src/riak_cs_s3_auth.erl [hunk 3]
```diff
@@ -130,43 +172,157 @@ calculate_signature(KeyData, RD) ->
 
     base64:encode_to_string(riak_cs_utils:sha_mac(KeyData, STS)).
 
+authenticate_v4(?RCS_USER{key_secret = SecretAccessKey} = _User, AuthAttrs, RD) ->
```
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:184]: I'd rather prefer having spec for this function to make code comprehensive.

**Comment by kuenishi** [comment]
I will add some eqc check at next sprint.

**Comment by kuenishi** [comment]
TODO: we have to add expiration and timestamp validation at `string_to_sign_v4/3` like v2, during next sprint.

File: src/riak_cs_s3_auth.erl [hunk 4]
```diff
@@ -130,43 +172,157 @@ calculate_signature(KeyData, RD) ->
...
+calculate_signature_v4(SecretAccessKey,
+                       {CredDate, AwsRegion, AwsService, AwsRequest}, StringToSign) ->
+    DateKey              = hmac_sha256(["AWS4", SecretAccessKey], CredDate),
+    DateRegionKey        = hmac_sha256(DateKey, AwsRegion),
+    DateRegionServiceKey = hmac_sha256(DateRegionKey, AwsService),
+    SigningKey           = hmac_sha256(DateRegionServiceKey, AwsRequest),
+    mochihex:to_hex(hmac_sha256(SigningKey, StringToSign)).
+
+hmac_sha256(Key, Data) ->
+    %% TODO: R15* compatibility?
```
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl]: Looks like this function could be moved to `riak_cs_utils` and put it under ifdefs. Or even they could be wrapped at webmachine? We can do it when we need it, later.

Comment (shino) [inline src/riak_cs_s3_auth.erl]: Do you think about R15B\* support? If so, R15B\* does not have sha256 things :crying_cat_face:

File: src/riak_cs_s3_auth.erl [hunk 5]
```diff
@@ -130,43 +172,157 @@ calculate_signature(KeyData, RD) ->
...
+    %% TODO: Expires?
+    TimeStamp = case lists:keyfind("x-amz-date", 1, AllHeaders) of
+                    false ->
+                        {"date", Date} = lists:keyfind("date", 1, AllHeaders),
+                        riak_cs_wm_utils:to_iso_8601(Date);
+                    {"x-amz-date", XAmzDate} ->
+                        XAmzDate
+                end,
+    {"Credential", Cred} = lists:keyfind("Credential", 1, AuthAttrs),
+    [_UserId, CredDate, AwsRegion, "s3" = AwsService, "aws4_request" = AwsRequest] =
```
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:246]: Let's add Region check later, at least after #1057 got merged.

Comment (shino) [inline src/riak_cs_s3_auth.erl:246]: I'm thinking whether it is good or bad. In other words, what good thing does region check bring for users?

File: src/riak_cs_s3_auth.erl [hunk 6]
```diff
@@ -130,43 +172,157 @@ calculate_signature(KeyData, RD) ->
...
+%% slashes (%2f).
+%% Because keys and values from raw query string that is URL-encoded at client
+%% side, they should be URL-decoded once and encoded back.
+strict_url_encode_for_qs_value(Value) ->
+    mochiweb_util:quote_plus(mochiweb_util:unquote(Value)).
+
+%% Force string URL encoding for path part of URL.
+%% Contrary to query part, slashes MUST NOT encoded and left as is.
+strict_url_encode_for_path(Path) ->
+    Tokens = binary:split(list_to_binary(Path), <<"/">>, [global]),
```
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:334]: Why did you choose `binary` but not `string:tokens/2`? Performance?

Comment (shino) [inline src/riak_cs_s3_auth.erl:334]: Not to drop preceding and trailing slashes.

```
5> binary:split(<<"/a/b/c/">>, <<"/">>, [global]).
[<<>>,<<"a">>,<<"b">>,<<"c">>,<<>>]
6> string:tokens("/a/b/c/", [$/]).
["a","b","c"]
```

File: src/riak_cs_s3_auth.erl (hunk 4 above)
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl]: Oh it doesn't. Cool. Let's update rebar.config.

Comment (kuenishi) [inline src/riak_cs_s3_auth.erl]: Mmm, or add OTP version check which considers `auth_v4_enabled` during the startup? Like if OTP is R15 and v4 enabled then it doesn't start.

Comment (shino) [inline src/riak_cs_s3_auth.erl]: Now Riak 2.0 supports only R16_. I will change rebar.config and not spend time for old R15_.

Comment (kuenishi) [inline src/riak_cs_s3_auth.erl]: :cool:

File: src/riak_cs_s3_auth.erl (hunk 5 above)
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:246]: S3 returns 400 when the region is wrong (s3cmd debug message):

```
DEBUG: Sending request method_string='GET', uri='/?delimiter=/', headers={'x-amz-content-sha256': 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b
7852b855', 'Authorization': 'AWS4-HMAC-SHA256 Credential=109QGDY2MTBVFVKMYT82/20150130/us-east-1/s3/aws4_request,SignedHeaders=host;x-amz-content-sha256;x-
amz-date,Signature=54bb86d974c176c862c1b444fe7c7503b81165b4bdfa40d4094d8db75461ce9b', 'x-amz-date': '20150130T054101Z'}, body=(0 bytes)
DEBUG: Response: {'status': 400, 'headers': {'x-amz-id-2': 'a/UhXeT7oFkC1zUokmGX/6hF0SabYJbcIMxQ6UF0oCIOFk7n2gN2U6Bn+0XuF3xL', 'server': 'AmazonS3', 'trans
fer-encoding': 'chunked', 'connection': 'close', 'x-amz-request-id': '9A79079BA0C545B5', 'date': 'Fri, 30 Jan 2015 05:42:21 GMT', 'content-type': 'applicat
ion/xml'}, 'reason': 'Bad Request', 'data': '<?xml version="1.0" encoding="UTF-8"?>\n<Error><Code>AuthorizationHeaderMalformed</Code><Message>The authoriza
tion header is malformed; the region \'us-east-1\' is wrong; expecting \'ap-northeast-1\'</Message><Region>ap-northeast-1</Region><RequestId>9A79079BA0C545
B5</RequestId><HostId>a/UhXeT7oFkC1zUokmGX/6hF0SabYJbcIMxQ6UF0oCIOFk7n2gN2U6Bn+0XuF3xL</HostId></Error>'}
DEBUG: ConnMan.put(): connection put back to pool (http://kuenishi.tokyo.s3.amazonaws.com#1)
```

s3cmd implicitly retries.

File: src/riak_cs_s3_auth.erl (hunk 6 above)
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:334]: :cool:

File: src/riak_cs_s3_auth.erl (hunk 5 above)
Comment (kuenishi) [inline src/riak_cs_s3_auth.erl:246]: Region verification does not improve security but just increases client-side configuration and lowers usability.

Comment (shino) [inline src/riak_cs_s3_auth.erl:246]: Yeah, it does such workaround.

So far, validating region does not improve security [1] and introduces some complexity for clients.
I will add comments but does not change code.

[1] For AWS, signing keys can be created by super-admin and normally one can use signing keys to access S3 and does not need to use his/her original-only-one secret key usually. Is it also nice for Riak CS and public service by CS? Hmm, maybe.

**Comment by shino** [comment]
Replaced by #1064.

**Pull request closed without merging**